File: Data Structures & Algorithms/lowest-common-ancestor-in-binary-search-tree/submission-4.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
        int high = std::max(p->val, q->val);
        int low = std::min(p->val, q->val);

        while (root) {
            if (root->val > high) {
                root = root->left;
            } else if (root->val < low) {
                root = root->right;
            } else {
                return root;
            }
        } 
        return nullptr;
    }
};

File: Data Structures & Algorithms/valid-binary-search-tree/submission-4.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        def helperIsValidBST(node, left, right):
            if not node:
                return True

            if not( left < node.val< right):
                return False

            return helperIsValidBST(node.left, left, node.val) and helperIsValidBST(node.right, node.val, right)
        
        return helperIsValidBST(root, -9999, 9999)
